File: examples/python/simple_server.py
"""
Simple MCP Server Example in Python

This demonstrates a basic MCP server that exposes both resources and tools.
It simulates a note-taking service with:
- Resources: Notes stored by the user
- Tools: Functions to create, update, and search notes
"""

import json
import asyncio
from datetime import datetime
from typing import Dict, List, Any, Optional

# Note: In a real implementation, you would use the actual MCP SDK
# from modelcontextprotocol import Server
# Since this is an example, we'll simulate the MCP server behavior

class NoteMCPServer:
    """A simulated MCP server for note-taking functionality."""
    
    def __init__(self):
        self.name = "NoteMCPServer"
        self.version = "1.0.0"
        self.notes = {
            "note-1": {
                "id": "note-1",
                "title": "Welcome to MCP Notes",
                "content": "This is your first note. You can create more using the create_note tool.",
                "tags": ["welcome", "introduction"],
                "created_at": "2025-05-01T10:00:00Z",
                "updated_at": "2025-05-01T10:00:00Z"
            },
            "note-2": {
                "id": "note-2",
                "title": "Shopping List",
                "content": "- Milk\n- Eggs\n- Bread\n- Apples",
                "tags": ["shopping", "groceries"],
                "created_at": "2025-05-02T14:30:00Z",
                "updated_at": "2025-05-02T14:30:00Z"
            }
        }
    
    async def list_resources(self) -> Dict[str, List[Dict[str, str]]]:
        """List available resources (notes)."""
        resources = []
        for note_id, note in self.notes.items():
            resources.append({
                "id": note_id,
                "name": note["title"],
                "description": f"Note created on {note['created_at'].split('T')[0]}"
            })
        return {"resources": resources}
    
    async def read_resource(self, resource_id: str) -> Dict[str, str]:
        """Read a specific resource (note)."""
        if resource_id in self.notes:
            return {"content": json.dumps(self.notes[resource_id], indent=2)}
        raise ValueError(f"Note {resource_id} not found")
    
    async def list_tools(self) -> Dict[str, List[Dict[str, Any]]]:
        """List available tools."""
        return {
            "tools": [
                {
                    "name": "create_note",
                    "description": "Create a new note",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "title": {
                                "type": "string",
                                "description": "Title of the note"
                            },
                            "content": {
                                "type": "string",
                                "description": "Content of the note"
                            },
                            "tags": {
                                "type": "array",
                                "items": {"type": "string"},
                                "description": "Tags for the note"
                            }
                        },
                        "required": ["title", "content"]
                    }
                },
                {
                    "name": "update_note",
                    "description": "Update an existing note",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "note_id": {
                                "type": "string",
                                "description": "ID of the note to update"
                            },
                            "title": {
                                "type": "string",
                                "description": "New title (optional)"
                            },
                            "content": {
                                "type": "string",
                                "description": "New content (optional)"
                            },
                            "tags": {
                                "type": "array",
                                "items": {"type": "string"},
                                "description": "New tags (optional)"
                            }
                        },
                        "required": ["note_id"]
                    }
                },
                {
                    "name": "search_notes",
                    "description": "Search notes by query or tags",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "query": {
                                "type": "string",
                                "description": "Search query (searches titles and content)"
                            },
                            "tags": {
                                "type": "array",
                                "items": {"type": "string"},
                                "description": "Filter by tags"
                            }
                        }
                    }
                }
            ]
        }
    
    async def call_tool(self, name: str, parameters: Dict[str, Any]) -> Dict[str, Any]:
        """Call a specific tool."""
        if name == "create_note":
            return await self._create_note(parameters)
        elif name == "update_note":
            return await self._update_note(parameters)
        elif name == "search_notes":
            return await self._search_notes(parameters)
        raise ValueError(f"Tool {name} not found")
    
    async def _create_note(self, parameters: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new note."""
        title = parameters["title"]
        content = parameters["content"]
        tags = parameters.get("tags", [])
        
        timestamp = datetime.utcnow().isoformat() + "Z"
        note_id = f"note-{len(self.notes) + 1}"
        
        self.notes[note_id] = {
            "id": note_id,
            "title": title,
            "content": content,
            "tags": tags,
            "created_at": timestamp,
            "updated_at": timestamp
        }
        
        return {
            "result": {
                "id": note_id,
                "title": title,
                "message": "Note created successfully"
            }
        }
    
    async def _update_note(self, parameters: Dict[str, Any]) -> Dict[str, Any]:
        """Update an existing note."""
        note_id = parameters["note_id"]
        
        if note_id not in self.notes:
            raise ValueError(f"Note {note_id} not found")
        
        note = self.notes[note_id]
        
        if "title" in parameters:
            note["title"] = parameters["title"]
        if "content" in parameters:
            note["content"] = parameters["content"]
        if "tags" in parameters:
            note["tags"] = parameters["tags"]
        
        note["updated_at"] = datetime.utcnow().isoformat() + "Z"
        
        return {
            "result": {
                "id": note_id,
                "title": note["title"],
                "message": "Note updated successfully"
            }
        }
    
    async def _search_notes(self, parameters: Dict[str, Any]) -> Dict[str, Any]:
        """Search notes by query or tags."""
        query = parameters.get("query", "").lower()
        tags = parameters.get("tags", [])
        
        results = []
        
        for note_id, note in self.notes.items():
            # Check if the note matches the query
            query_match = (
                not query or
                query in note["title"].lower() or
                query in note["content"].lower()
            )
            
            # Check if the note has all the required tags
            tag_match = (
                not tags or
                all(tag in note["tags"] for tag in tags)
            )
            
            if query_match and tag_match:
                results.append({
                    "id": note_id,
                    "title": note["title"],
                    "preview": note["content"][:100] + "..." if len(note["content"]) > 100 else note["content"],
                    "tags": note["tags"],
                    "updated_at": note["updated_at"]
                })
        
        return {
            "result": {
                "count": len(results),
                "notes": results
            }
        }
    
    async def process_request(self, request: Dict[str, Any]) -> Dict[str, Any]:
        """Process an MCP JSON-RPC request."""
        method = request.get("method")
        params = request.get("params", {})
        
        if method == "ListResources":
            return await self.list_resources()
        elif method == "ReadResource":
            return await self.read_resource(params["resourceId"])
        elif method == "ListTools":
            return await self.list_tools()
        elif method == "CallTool":
            return await self.call_tool(params["name"], params["parameters"])
        else:
            raise ValueError(f"Unsupported method: {method}")

    async def listen(self):
        """Simulates the MCP server listening for requests."""
        print(f"{self.name} v{self.version} is running...")
        print("This is a simulated server. In a real implementation, you would use:")
        print("  from modelcontextprotocol import Server")
        print("  server = Server(name='NoteMCPServer', version='1.0.0', features={...})")
        print("  @server.list_resources")
        print("  async def list_resources():")
        print("      # Your implementation")
        print("  server.listen()")

async def main():
    server = NoteMCPServer()
    await server.listen()
    
    # Example of how client interaction would work
    print("\nSimulating client requests:")
    
    # List resources
    print("\n> Client requests ListResources")
    response = await server.process_request({"method": "ListResources"})
    print(f"< Server responds: {json.dumps(response, indent=2)}")
    
    # Create a note
    print("\n> Client calls CreateNote tool")
    response = await server.process_request({
        "method": "CallTool",
        "params": {
            "name": "create_note",
            "parameters": {
                "title": "Meeting Notes",
                "content": "Discussed MCP integration plans for Q3",
                "tags": ["meeting", "planning"]
            }
        }
    })
    print(f"< Server responds: {json.dumps(response, indent=2)}")
    
    # Search notes
    print("\n> Client searches for 'meeting' notes")
    response = await server.process_request({
        "method": "CallTool",
        "params": {
            "name": "search_notes",
            "parameters": {
                "tags": ["meeting"]
            }
        }
    })
    print(f"< Server responds: {json.dumps(response, indent=2)}")

if __name__ == "__main__":
    asyncio.run(main()) 

File: examples/typescript/simple-server.ts
/**
 * Simple MCP Server Example
 * 
 * This demonstrates a basic MCP server that exposes both resources and tools.
 * It simulates a weather data service with:
 * - Resources: Current weather data for different cities
 * - Tools: Functions to convert temperatures and get forecasts
 */

import { Server, ServerOptions } from '@modelcontextprotocol/sdk';

// Initialize the server
const options: ServerOptions = {
  name: 'WeatherMCPServer',
  version: '1.0.0',
  features: {
    resources: true,
    tools: true,
    prompts: false
  }
};

const server = new Server(options);

// Simulated weather data
const weatherData = {
  'new-york': {
    city: 'New York',
    temperature: 72,
    condition: 'Partly Cloudy',
    humidity: 65,
    updatedAt: new Date().toISOString()
  },
  'london': {
    city: 'London',
    temperature: 16,
    condition: 'Rainy',
    humidity: 80,
    updatedAt: new Date().toISOString()
  },
  'tokyo': {
    city: 'Tokyo',
    temperature: 25,
    condition: 'Clear',
    humidity: 60,
    updatedAt: new Date().toISOString()
  }
};

// Register resource handlers
server.registerResourceHandler({
  // List available resources (cities)
  listResources: async () => {
    return {
      resources: Object.keys(weatherData).map(cityId => ({
        id: cityId,
        name: weatherData[cityId].city,
        description: `Current weather for ${weatherData[cityId].city}`
      }))
    };
  },
  
  // Read a specific resource (city weather data)
  readResource: async ({ resourceId }) => {
    if (resourceId in weatherData) {
      return {
        content: JSON.stringify(weatherData[resourceId], null, 2)
      };
    }
    
    throw new Error(`Weather data for ${resourceId} not found`);
  }
});

// Register tool handlers
server.registerToolHandler({
  // List available tools
  listTools: async () => {
    return {
      tools: [
        {
          name: 'convert_temperature',
          description: 'Convert temperature between Fahrenheit and Celsius',
          parameters: {
            type: 'object',
            properties: {
              temperature: {
                type: 'number',
                description: 'Temperature value to convert'
              },
              from_unit: {
                type: 'string',
                enum: ['celsius', 'fahrenheit'],
                description: 'Current temperature unit'
              },
              to_unit: {
                type: 'string',
                enum: ['celsius', 'fahrenheit'],
                description: 'Target temperature unit'
              }
            },
            required: ['temperature', 'from_unit', 'to_unit']
          }
        },
        {
          name: 'get_forecast',
          description: 'Get weather forecast for a city',
          parameters: {
            type: 'object',
            properties: {
              city: {
                type: 'string',
                description: 'City name'
              },
              days: {
                type: 'number',
                description: 'Number of days for forecast (1-7)'
              }
            },
            required: ['city']
          }
        }
      ]
    };
  },
  
  // Call a specific tool
  callTool: async ({ name, parameters }) => {
    if (name === 'convert_temperature') {
      const { temperature, from_unit, to_unit } = parameters as { 
        temperature: number; 
        from_unit: 'celsius' | 'fahrenheit'; 
        to_unit: 'celsius' | 'fahrenheit' 
      };
      
      if (from_unit === to_unit) {
        return { result: temperature };
      }
      
      let result: number;
      if (from_unit === 'celsius' && to_unit === 'fahrenheit') {
        result = (temperature * 9/5) + 32;
      } else {
        result = (temperature - 32) * 5/9;
      }
      
      return {
        result: {
          value: parseFloat(result.toFixed(1)),
          unit: to_unit
        }
      };
    }
    
    if (name === 'get_forecast') {
      const { city, days = 3 } = parameters as { city: string; days?: number };
      const normalizedCity = city.toLowerCase().replace(/\s+/g, '-');
      
      // Check if we have data for this city
      const cityData = Object.values(weatherData).find(
        data => data.city.toLowerCase() === city.toLowerCase()
      );
      
      if (!cityData) {
        throw new Error(`No weather data available for ${city}`);
      }
      
      // Generate a simulated forecast
      const forecast = [];
      const conditions = ['Sunny', 'Partly Cloudy', 'Cloudy', 'Rainy', 'Stormy', 'Clear'];
      const baseTemp = cityData.temperature;
      
      for (let i = 1; i <= Math.min(days, 7); i++) {
        const variation = Math.floor(Math.random() * 10) - 5; // -5 to +5 degree variation
        forecast.push({
          day: i,
          date: new Date(Date.now() + i * 86400000).toISOString().split('T')[0],
          temperature: baseTemp + variation,
          condition: conditions[Math.floor(Math.random() * conditions.length)],
        });
      }
      
      return {
        result: {
          city: cityData.city,
          forecast
        }
      };
    }
    
    throw new Error(`Tool ${name} not found`);
  }
});

// Start the server
server.listen();

console.log('Weather MCP Server is running...'); 